src/lib/mongoBackend/mongoSubCache.cpp

```diff
@@ -0,0 +1,1204 @@
+
+using namespace mongo;
+using std::auto_ptr;
+
+
+
+/* ****************************************************************************
+*
+* EntityInfo::EntityInfo - 
+*/
+EntityInfo::EntityInfo(const std::string& idPattern, const std::string& _entityType)
+{
+  regcomp(&entityIdPattern, idPattern.c_str(), 0);
+
+  entityType               = _entityType;
+  entityIdPatternToBeFreed = true;
+}
+
+
+
+/* ****************************************************************************
+*
+* EntityInfo::match - 
+*/
+bool EntityInfo::match
+(
+  const std::string&  id,
+  const std::string&  type
+)
+{
+  //
+  // If type non-empty - perfect match is mandatory
+  // If type is empty - always matches
+  //
+  if ((type != "") && (entityType != type) && (entityType != ""))
+  {
+    return false;
+  }
+
+  // REGEX-comparison this->entityIdPattern VS id
+  return regexec(&entityIdPattern, id.c_str(), 0, NULL, 0) == 0;
+}
+
+
+
+/* ****************************************************************************
+*
+* EntityInfo::release - 
+*/
+void EntityInfo::release(void)
+{
+  if (entityIdPatternToBeFreed)
+  {
+    regfree(&entityIdPattern);
+    entityIdPatternToBeFreed = false;
+  }
+}
+
+
+
+/* ****************************************************************************
+*
+* EntityInfo::present - 
+*/
+void EntityInfo::present(const std::string& prefix)
+{
+  LM_F(("%sid:    regex describing EntityId::id (the original string is not saved)", prefix.c_str()));
+  LM_F(("%stype:  %s", prefix.c_str(), entityType.c_str()));
+}
+
+
+
+/* ****************************************************************************
+*
+* MongoSubCache - 
+*/
+typedef struct MongoSubCache
+{
+  CachedSubscription* head;
+  CachedSubscription* tail;
+
+  // Statistics counters
+  int                 noOfRefreshes;
+  int                 noOfInserts;
+  int                 noOfRemoves;
+  int                 noOfUpdates;
+} MongoSubCache;
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCache - 
+*/
+MongoSubCache  mongoSubCache = { NULL, NULL, 0 };
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheInit - 
+*/
+void mongoSubCacheInit(void)
+{
+  LM_T(LmtMongoSubCache, ("Initializing subscription cache"));
+
+  mongoSubCache.head   = NULL;
+  mongoSubCache.tail   = NULL;
+
+  mongoSubCacheStatisticsReset("mongoSubCacheInit");
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItems - 
+*/
+int mongoSubCacheItems(void)
+{
+  CachedSubscription* cSubP = mongoSubCache.head;
+  int                 items = 0;
+
+  while (cSubP != NULL)
+  {
+    ++items;
+    cSubP = cSubP->next;
+  }
+
+  return items;
+}
+
+
+
+/* ****************************************************************************
+*
+* mongoSubCacheItemInsert - 
+*/
+static void mongoSubCacheItemInsert(CachedSubscription* cSubP)
```

Comment:
Maybe the implementation of this method should be closer to the other instances of mongoSubCacheItemInsert() methods.